### DISCUSSION TITLE
kovplus game not work via mame emscripten

### DISCUSSION THREAD

**Discussion by liudonghua123** [mamedev/discussions#8] [General]
Hi, I tried to build an emscripten build of mame 0.246 in order to run some games in the browser. I encountered some problems when build the mame https://github.com/mamedev/mame/issues/10257, and finally I build successfully.

I use `emmake make SUBTARGET=arcade SOURCES=src/mame/neogeo/neogeo.cpp TOOLS=0 NOWERROR=1 REGENIE=1 -j8` to build the wasm/js/html. And I also tried `emmake make SUBTARGET=neogeo SOURCES=src/mame/neogeo/neogeo.cpp TOOLS=0 NOWERROR=1 REGENIE=1 -j8`, both of them produce the same files except filename different. 

Then I use https://github.com/db48x/emularity/ to load the game. The configuration part of code is the following.

```js
      var emulator = new Emulator(
        document.querySelector("#canvas"),
        null,
        new JSMAMELoader(
          JSMAMELoader.driver("neogeo"),
          JSMAMELoader.nativeResolution(256, 256),
          JSMAMELoader.scale(3),
          JSMAMELoader.emulatorJS("emulators/mame/neogeo.js"),
          JSMAMELoader.mountFile(
            "pgm.zip",
            JSMAMELoader.fetchFile("Game File", "emulators/roms/pgm.zip")
          ),
          JSMAMELoader.mountFile(
            "neogeo.zip",
            JSMAMELoader.fetchFile("Game File", "emulators/roms/neogeo.zip")
          ),
          JSMAMELoader.mountFile(
            "kovplus.zip",
            JSMAMELoader.fetchFile("Game File", "emulators/roms/kovplus.zip")
          )
        )
      );
      emulator.start({ waitAfterDownloading: true });
```

No errors on the console of devtools. I could not find any clues to continue to promote the work. Does anyone can help me. 

Thanks very much in advance.

![image](https://user-images.githubusercontent.com/2276718/186295643-a3eb975b-6a81-4519-a417-a0f20363116e.png)

I also tested some other games like `tengai` (`emmake make SUBTARGET=psikyo SOURCES=src/mame/psikyo/psikyo.cpp -j8 NOWERROR=1`), `shadfrce` (`emmake make SUBTARGET=shadfrce SOURCES=src/mame/technos/shadfrce.cpp -j8 NOWERROR=1`). Both of the two works.

```js
        // the configuration part of code of tengai
        new JSMAMELoader(
          JSMAMELoader.driver("tengai"),
          JSMAMELoader.nativeResolution(256, 256),
          JSMAMELoader.scale(3),
          JSMAMELoader.emulatorJS("emulators/mame/psikyo.js"),
          JSMAMELoader.mountFile(
            "tengai.zip",
            JSMAMELoader.fetchFile("Game File", "emulators/roms/tengai.zip")
          )
        )

        // the configuration part of code of shadfrce
        new JSMAMELoader(
          JSMAMELoader.driver("shadfrce"),
          JSMAMELoader.nativeResolution(256, 256),
          JSMAMELoader.scale(3),
          JSMAMELoader.emulatorJS("emulators/mame/shadfrce.js"),
          JSMAMELoader.mountFile(
            "shadfrce.zip",
            JSMAMELoader.fetchFile("Game File", "emulators/roms/shadfrce.zip")
          )
        )

**Comment by Tafoid** [comment]
I think you need to re-examine which driver you are compiling to use for **kovplus** - that machine is not a NeoGeo game.  In essence you are trying to insert a game in with a driver it is not designed for.. unless I'm missing something less obvious.

**Reply by liudonghua123** [reply]
`kovplus` is not a neogeo game? Oh, yes, I found it maybe a `pgm` game according to [kovplus](https://www.planetemu.net/rom/mame-roms/kovplus). Thanks, I will try to build the `pgm` driver and test again.

**Reply by liudonghua123** [reply]
Thanks, I tried to use `emmake make SUBTARGET=pgm SOURCES=src/mame/igs/pgm.cpp -j8 NOWERROR=1` to build and use the following code, then it worked. Except the game is too stutteringly.

The driver of the support game is listed in https://raw.githubusercontent.com/mamedev/mame/master/src/mame/mame.lst.

https://user-images.githubusercontent.com/2276718/186304860-e9883a06-a876-48ed-980f-c1ef8dd2b3c7.mp4

```js
      var emulator = new Emulator(
        document.querySelector("#canvas"),
        null,
        new JSMAMELoader(
          JSMAMELoader.driver("kovplus"),
          JSMAMELoader.nativeResolution(256, 256),
          JSMAMELoader.scale(3),
          JSMAMELoader.emulatorJS("emulators/mame/pgm.js"),
          JSMAMELoader.mountFile(
            "pgm.zip",
            JSMAMELoader.fetchFile("Bios", "emulators/roms/pgm.zip")
          ),
          JSMAMELoader.mountFile(
            "kovplus.zip",
            JSMAMELoader.fetchFile("Game File", "emulators/roms/kovplus.zip")
          )
        )
      );
      emulator.start({ waitAfterDownloading: true });
```

I tried [emscripten build of RetroArch](https://buildbot.libretro.com/stable/1.10.3/emscripten/RetroArch.7z) a few days ago. and use the `mame 2003 plus` kernel to run `kovplus` game, which is normal and playable.

Is there any optimization guidence about emscripten build of mame.

**Comment by liudonghua123** [comment]
I also tried another `neogeo` game `mslug`. I changed the driver `neogeo` to `mslug` (see https://github.com/db48x/emularity/issues/57), then it worked. But it did not work for `kovplus` game.

```js

        new JSMAMELoader(
          JSMAMELoader.driver("mslug"),
          JSMAMELoader.nativeResolution(256, 256),
          JSMAMELoader.scale(3),
          JSMAMELoader.emulatorJS("emulators/mame/neogeo.js"),
          JSMAMELoader.mountFile(
            "neogeo.zip",
            JSMAMELoader.fetchFile("Bios", "emulators/roms/neogeo.zip")
          ),
          JSMAMELoader.mountFile(
            "mslug.zip",
            JSMAMELoader.fetchFile("Game File", "emulators/roms/mslug.zip")
          )
        )
```

**Comment by liudonghua123** [comment]
Btw, how can I run the generated HTML, how to configure the bios/rom like [emularity](https://github.com/db48x/emularity). I have build the `common.{js,wasm,html}` via `emmake make SUBTARGET=common SOURCES=igs/pgm.cpp,neogeo/neogeo.cpp,irem/m92.cpp,psikyo/psikyo.cpp,technos/shadfrce.cpp,atari/batman.cpp -j8 NOWERROR=1`. I got the following errors when I tried to run some games. I use [`serve`](https://github.com/vercel/serve) cli to host these files on localhost, I even copied these files to the directory which contains mame executable with other directory or files needed by mame executable.

![image](https://user-images.githubusercontent.com/2276718/186347291-32393ea9-0a96-43ed-8957-e7bcb1de77ed.png)
